Mathlib/Geometry/Manifold/IntegralCurve.lean

```diff
@@ -0,0 +1,767 @@
+  · rw [mem_Ioo]
+    exact ⟨neg_lt_zero.mpr hpos, by positivity⟩
+  · apply Ioo_subset_Ioo <;> linarith
+
+/-- Auxiliary lemma. Suppose for every `a : ℝ`, there exists an integral curve defined on
+  `Ioo (-a) a`. We choose one for every `a` and call it `γ a`. We define a global curve
+  `γ_ext := fun t ↦ γ (|t| + 1) t`. For every `a : ℝ`, `γ` agrees with `γ a` on `Ioo (-a) a`. This
+  will help us show that `γ` is a global integral curve. -/
+lemma integralCurve_of_exists_isIntegralCurveOn_Ioo_eqOn [BoundarylessManifold I M]
+    (hv : ContMDiff I I.tangent 1 (fun x => (⟨x, v x⟩ : TangentBundle I M))) {x : M}
+    (h : ∀ a, ∃ γ, γ 0 = x ∧ IsIntegralCurveOn γ v (Ioo (-a) a)) {a : ℝ} :
+    EqOn (fun t' ↦ choose (h (|t'| + 1)) t') (choose (h a)) (Ioo (-a) a) := by
+  intros t' ht'
+  by_cases hlt : |t'| + 1 < a
+  · apply exists_isIntegralCurveOn_Ioo_eqOn hv h (by positivity) (le_of_lt hlt)
+    · rw [mem_Ioo, ← abs_lt]
+      exact lt_add_one _
+  · apply (exists_isIntegralCurveOn_Ioo_eqOn hv h _ _ _).symm
+    · have := lt_trans ht'.1 ht'.2
+      rw [neg_lt_self_iff] at this
+      exact this
+    · exact not_lt.mp hlt
+    · exact ht'
+
+/-- If for every `a : ℝ`, there exists an integral curve defined on `Ioo (-a) a`, then there exists
+  a global integral curve. -/
+lemma exists_integralCurve_of_exists_isIntegralCurveOn_Ioo [BoundarylessManifold I M] [T2Space M]
+    (hv : ContMDiff I I.tangent 1 (fun x => (⟨x, v x⟩ : TangentBundle I M))) {x : M}
+    (h : ∀ a, ∃ γ, γ 0 = x ∧ IsIntegralCurveOn γ v (Ioo (-a) a)) :
+    ∃ γ, γ 0 = x ∧ IsIntegralCurve γ v := by
+  refine ⟨fun t' ↦ choose (h (|t'| + 1)) t', (choose_spec (h (|0| + 1))).1, ?_⟩
+  intro t
+  apply HasMFDerivAt.congr_of_eventuallyEq (f := choose (h (|t| + 1)))
+  · apply (choose_spec (h (|t| + 1))).2 t
+    rw [mem_Ioo, ← abs_lt]
+    exact lt_add_one _
+  · rw [Filter.eventuallyEq_iff_exists_mem]
+    refine ⟨Ioo (-(|t| + 1)) (|t| + 1), ?_, integralCurve_of_exists_isIntegralCurveOn_Ioo_eqOn hv h⟩
+    · have := lt_add_of_pos_right |t| zero_lt_one
+      rw [abs_lt] at this
+      exact Ioo_mem_nhds this.1 this.2
+
+lemma exists_isIntegralCurve_iff_exists_isIntegralCurveOn_Ioo [BoundarylessManifold I M] [T2Space M]
+    (hv : ContMDiff I I.tangent 1 (fun x => (⟨x, v x⟩ : TangentBundle I M))) (x : M) :
+    (∃ γ, γ 0 = x ∧ IsIntegralCurve γ v) ↔
+      ∀ a, ∃ γ, γ 0 = x ∧ IsIntegralCurveOn γ v (Ioo (-a) a) :=
+  ⟨fun ⟨γ, h1, h2⟩ _ ↦ ⟨γ, h1, h2.isIntegralCurveOn _⟩,
+   exists_integralCurve_of_exists_isIntegralCurveOn_Ioo hv⟩
+
+lemma piecewise_eqOn_symm [BoundarylessManifold I M]
+    (hv : ContMDiff I I.tangent 1 (fun x => (⟨x, v x⟩ : TangentBundle I M)))
+    {a b a' b' : ℝ} (hγ : IsIntegralCurveOn γ v (Ioo a b))
+    (hγ' : IsIntegralCurveOn γ' v (Ioo a' b'))
+    (ht₀ : t₀ ∈ Ioo a b ∩ Ioo a' b') (h : γ t₀ = γ' t₀) :
+    EqOn (piecewise (Ioo a b) γ γ') γ' (Ioo a' b') := by
+  intros t ht
+  by_cases hle : t ≤ a
+  · rw [piecewise, if_neg]
```

Comment:
This argument is present twice and can surely be factored out.

Reply:
Golfed!